packages/tools/viewer-alpha/src/viewer.ts

```diff
@@ -0,0 +1,240 @@
+        this._camera = new ArcRotateCamera("camera1", 0, 0, 1, Vector3.Zero(), this._scene);
+        this._camera.attachControl();
+        this._reframeCamera(); // set default camera values
+
+        // TODO: render at least back ground. Maybe we can only run renderloop when a mesh is loaded. What to render until then?
+        this._engine.runRenderLoop(() => {
+            this._scene.render();
+        });
+
+        options?.onInitialized?.({ scene: this._scene, camera: this._camera });
+    }
+
+    /**
+     * Loads a 3D model from the specified URL.
+     * @remarks
+     * If a model is already loaded, it will be unloaded before loading the new model.
+     * @param url The URL of the model to load.
+     * @param abortSignal An optional signal that can be used to abort the loading process.
+     */
+    public async loadModelAsync(url: string, abortSignal?: AbortSignal): Promise<void> {
+        if (this._isDisposed) {
+            throw new Error("Viewer is disposed");
+        }
+
+        this._loadModelAbortController?.abort();
+        const abortController = (this._loadModelAbortController = new AbortController());
+
+        const throwIfAborted = () => {
+            // External cancellation
+            abortSignal?.throwIfAborted();
+
+            // Internal cancellation
+            abortController.signal.throwIfAborted();
+        };
+
+        await this._loadModelLock.lockAsync(async () => {
+            throwIfAborted();
+            this._assetContainer?.dispose();
+            this._assetContainer = await SceneLoader.LoadAssetContainerAsync("", url, this._scene);
+            this._assetContainer.addAllToScene();
+            this._reframeCamera();
+        });
+    }
+
+    /**
+     * Loads an environment texture from the specified URL and sets up a corresponding skybox.
+     * @remarks
+     * If no URL is provided, a default hemispheric light will be created.
+     * If an environment is already loaded, it will be unloaded before loading the new environment.
+     * @param url The URL of the environment texture to load.
+     * @param abortSignal An optional signal that can be used to abort the loading process.
+     */
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    public async loadEnvironmentAsync(url: Nullable<string | undefined>, abortSignal?: AbortSignal): Promise<void> {
+        if (this._isDisposed) {
+            throw new Error("Viewer is disposed");
+        }
+
+        this._loadEnvironmentAbortController?.abort();
+        const abortController = (this._loadEnvironmentAbortController = new AbortController());
+
+        const throwIfAborted = () => {
+            // External cancellation
+            abortController.signal.throwIfAborted();
+
+            // Internal cancellation
+            abortController.signal.throwIfAborted();
+        };
+
+        await this._loadEnvironmentLock.lockAsync(async () => {
+            throwIfAborted();

```

Comment:
Do you need to be checking _isDisposed inside each of these locked blocks (perhaps in throwIfAborted?) as well as after each await inside each of these?  I've never liked having to check state after each await - would things just unwind cleanly in the case that dispose is called one of these async operations?

Reply:
No it would not "unwind cleanly" as many of the Babylon objects themselves don't check their disposed state. You are right that I should have a dispose check here. Will update.

Reply:
It'd be great to make this pattern fail-safe (hard to forget to check isDisposed everywhere the check is needed - e.g. after each await).  I haven't seen a great general solution to this problem - any ideas?  Maybe an await wrapper?

Reply:
It's a little tricky because if it is being executed synchronously as a call from the consumer of the API, then we should throw for sure. If the dispose happens during an async operation, then maybe we should still throw, but possibly a different error, more like "this async operation was implicitly cancelled as the result of the object being disposed."

The async operation also could be from a promise returning async function, or it could be in the form of a callback, which also complicates having a general purpose solution. 

If we only cared about promise returning async functions, then maybe we could have a higher level wrapper that can execute multiple behaviors so you don't end up wrapping async functions in multiple layers (such as my usage of AsyncLock in this file). Maybe something like `wrapAsync(someAsyncFunction, lockBehavior, disposeCheckBehavior, ...)`.

We could consider something like this more, but for now I will just have a local helper for the patterns present in this file.